File: projetoJava/src/Produto.java

public class Produto {
    private String nome;
    private double preco;

    public Produto(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    public Produto somar(Produto outro) {
        String novoNome = this.nome + " e " + outro.nome;
        double novoPreco = this.preco + outro.preco;
        return new Produto(novoNome, novoPreco);
    }

    @Override
    public String toString() {
        return nome + ": R$ " + String.format("%.2f", preco);
    }
}

File: projetoGo/produto.go
package main

import "fmt"

type Produto struct {
	nome  string
	preco float64
}

func NovoProduto(nome string, preco float64) Produto {
	return Produto{nome: nome, preco: preco}
}

func (p Produto) Somar(outro Produto) Produto {
	novoNome := fmt.Sprintf("%s e %s", p.nome, outro.nome)
	novoPreco := p.preco + outro.preco
	return NovoProduto(novoNome, novoPreco)
}

func (p Produto) String() string {
	return fmt.Sprintf("%s: R$ %.2f", p.nome, p.preco)
}
